dfusion_rust_core/src/models/solution.rs

```diff
@@ -17,12 +17,78 @@ pub struct Solution {
 impl Solution {
     pub fn trivial(num_orders: usize) -> Self {
         Solution {
-            objective_value: Some(U256::zero()),
             prices: vec![0; TOKENS as usize],
             executed_buy_amounts: vec![0; num_orders],
             executed_sell_amounts: vec![0; num_orders],
         }
     }
+
+    /// Returns the token price for token or `None` if the token is invalid.
+    fn price(&self, token: u16) -> Option<u128> {
+        self.prices.get(token as usize).cloned()
+    }
+
+    /// Compute the result of the objective function for the solution. This is
+    /// currently defined as *U - dUT* where *U* is the total utility of the
+    /// trade and *dUT* is the total disreguarded utility of touched orders.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` for any solution that is positive valid solution.
+    /// Returns `None` for invalid solutions or orders. A solution might be
+    /// invalid if:
+    /// - The number of orders do not match the solution buy/sell amounts
+    /// - The solution values lead to invalid utility or disregarded utility
+    ///   results
+    /// - The orders reference tokens that are not priced by the solution
+    ///
+    /// Note that just because this method returns some value does not mean that
+    /// the solution is actually valid WRT to the smart contract as it currently
+    /// does not verify that things like `execSellAmount * p[sellToken] * (1-phi)
+    /// == execBuyAmount * p[buyToken]` or token conservation hold.
+    pub fn objective_value(&self, orders: &[Order]) -> Option<U256> {
+        if orders.len() != self.executed_buy_amounts.len()
+            || orders.len() != self.executed_sell_amounts.len()
+        {
+            return None;
+        }
+
+        // note that we don't actually have to do the same checking here (such as
+        // account balances and such) as we do in the smart contract - we only
+        // care if the solution is valid given the provided orders
+
+        let mut u = U256::zero();
+        let mut du = U256::zero();
+        let mut fee = I256::zero();
```

Comment:
Maybe we don't even need the I256.

If we would run two times through the orders, and first do all the feeToken balance adjustments for the cases where the feeToken is bought, and then, in a second loop, doing all the adjustments for the cases where the feeToken is sold, the fee should never be negative.

And later for calculating the objective value, we would just add and substract web3::U256 types with checks that we will never get negative. ( If we decided for this strategy, you check the smart contracts. They are already implementing this).

I think, in the near future, we don't need the I256, and hence, we could try to do it without. 
But for sure, for simplicity, future-readiness and readability of code etc, we could go with this variant.

Reply:
That's a great idea! We can leave the `I256` in a branch somewhere if we want it in the future.